src/core/lib/channel/channel_tracer.c

```diff
@@ -0,0 +1,325 @@
+  return tracer;
+}
+#endif
+
+static void free_node(grpc_trace_node* node) {
+  // no need to free string, since they are always static
+  GRPC_ERROR_UNREF(node->error);
+  if (node->subchannel) {
+    GRPC_CHANNEL_TRACER_UNREF(node->subchannel);
+  }
+  gpr_free(node);
+}
+
+void grpc_channel_tracer_destroy(grpc_channel_tracer* tracer) {
+  grpc_trace_node* it = tracer->node_list.head_trace;
+  while (it) {
+    grpc_trace_node* to_free = it;
+    it = it->next;
+    free_node(to_free);
+  }
+  gpr_free(tracer);
+}
+
+#ifdef GRPC_CHANNEL_TRACER_REFCOUNT_DEBUG
+void grpc_channel_tracer_unref(grpc_channel_tracer* tracer, const char* file,
+                               int line, const char* func) {
+  if (!tracer) return;
+  gpr_log(GPR_DEBUG, "%p: %" PRIdPTR " -> %" PRIdPTR " [%s:%d %s]", tracer,
+          gpr_atm_no_barrier_load(&tracer->refs.count),
+          gpr_atm_no_barrier_load(&tracer->refs.count) - 1, file, line, func);
+  if (gpr_unref(&tracer->refs)) {
+    grpc_channel_tracer_destroy(tracer);
+  }
+}
+#else
+void grpc_channel_tracer_unref(grpc_channel_tracer* tracer) {
+  if (!tracer) return;
+  if (gpr_unref(&tracer->refs)) {
+    grpc_channel_tracer_destroy(tracer);
+  }
+}
+#endif
+
+static void add_trace(grpc_trace_node_list* list, const char* trace,
+                      grpc_error* error,
+                      grpc_connectivity_state connectivity_state,
+                      grpc_channel_tracer* subchannel) {
+  grpc_trace_node* new_trace_node = gpr_malloc(sizeof(grpc_trace_node));
+  new_trace_node->data = trace;
+  new_trace_node->error = error;
+  new_trace_node->time_created = gpr_now(GPR_CLOCK_REALTIME);
+  new_trace_node->connectivity_state = connectivity_state;
+  new_trace_node->next = NULL;
+  new_trace_node->subchannel = subchannel;
+
+  if (subchannel) {
+    GRPC_CHANNEL_TRACER_REF(subchannel);
+  }
+
+  // first node in case
+  if (!list->head_trace) {
+    list->head_trace = list->tail_trace = new_trace_node;
+  }
+  // regular node add case
+  else {
+    list->tail_trace->next = new_trace_node;
+    list->tail_trace = list->tail_trace->next;
+  }
+  list->size++;
+
+  // maybe garbage collect the end
+  if (list->size > list->max_size) {
+    grpc_trace_node* to_free = list->head_trace;
+    list->head_trace = list->head_trace->next;
+    free_node(to_free);
+    list->size--;
+  }
+}
+
+void grpc_channel_tracer_add_trace(grpc_channel_tracer* tracer,
+                                   const char* trace, grpc_error* error,
+                                   grpc_connectivity_state connectivity_state,
+                                   grpc_channel_tracer* subchannel) {
+  if (!tracer) return;
+  tracer->num_nodes_logged++;
+  add_trace(&tracer->node_list, trace, error, connectivity_state, subchannel);
+}
+
+// TODO(ncteisen): pull this function into a helper location
+static grpc_json* link_json(grpc_json* child, grpc_json* brother,
+                            grpc_json* parent) {
```

Comment:
(1) What's the advantage of creating a json tree here (versus just generating a string)?

(2) Given we're already thinking about a proto format for this, is it worth pulling the presentation of traces (the json generation) into a utility file so that we don't accidentally link presentation with runtime storage?